rclcpp/src/rclcpp/executors/multi_threaded_executor.cpp

```diff
@@ -56,17 +74,35 @@ MultiThreadedExecutor::spin()
     throw std::runtime_error("spin() called while already spinning");
   }
   RCPPUTILS_SCOPE_EXIT(this->spinning.store(false); );
-  std::vector<std::thread> threads;
+  std::vector<rcpputils::Thread> threads;
   size_t thread_id = 0;
-  {
-    std::lock_guard wait_lock{wait_mutex_};
-    for (; thread_id < number_of_threads_ - 1; ++thread_id) {
-      auto func = std::bind(&MultiThreadedExecutor::run, this, thread_id);
-      threads.emplace_back(func);
+
+  if (thread_attributes_) {
+    rcpputils::Thread::Attribute thread_attr;
+    {
+      std::lock_guard wait_lock{wait_mutex_};
+      for (; thread_id < thread_attributes_->num_attributes - 1; ++thread_id) {
+        thread_attr.set_thread_attribute(
+          thread_attributes_->attributes[thread_id]);
```

Comment:
The Multi-Threaded Executor has multiple threads to process the callback for new incoming messages. It does not differentiate between different messages or timers. 

What is your intention to assign different thread parameters (e.g. priority, core-affinity) to different threads? At run-time you cannot control which callback is processed by which thread. 

Could you give an example how you intend to use the multi-threaded Executor with different thread priorities in your use-case?

Reply:
Thanks for your effort to add thread configuration to ROS2. This is very important piece of work regarding real-time support. I think we need a proper way to control the thread configuration of the thread, which is processing a callback. I am not so sure, that directly assigning the priority in the Multi-Threaded Executor solves the problem. You also need to make a mapping of callbacks to threads. 

Other approaches are to create a thread *outside* ROS 2 and then let e.g. the single threaded Executor run. 

Another approach is the change the thread priority just before the callback is called. This is proposed in an experimental real-time executor at the "Real-time programming with ROS 2" workshop at ROSCon. The source code will be made available.

Reply:
> Other approaches are to create a thread outside ROS 2 and then let e.g. the single threaded Executor run.

I agree with this.
We should not have executors modify the attributes of the thread that they are running into. If users want to run a single-threaded executor in a thread with specific attributes, they can easily do it themselves, rather than having the executor change things under their feet.

A different scenario is whenever the rclcpp library internally creates threads (e.g. the events-executor, the clock subscriptions, the graph updates etc) all these can take advantage of the new tools to configure threads (rather than just blindly inheriting the parent thread configurations).


Besides that, I'm not convinced by the use of
```
rcutils_thread_attrs_t * attrs = rcl_context_get_thread_attrs(options.context->get_rcl_context().get()
```

We shouldn't restrict to a single thread attribute per ROS context and the two concepts shouldn't be tied together.
I would recommend to have all classes that deal with threads to take a `rcutils_thread_attrs_t` as input in the constructor